Unikernel Deployment (#16)
nitindermohan/oak-new-docs: content/docs/manuals/unikernel-operations.md

---
title: "Unikernel Deployment"
summary: ""
draft: false
weight: 301
toc: true
seo:
  title: "Unikernel Deployment" # custom title (optional)
  description: "Deployment of Unikraft Unikernels in Oakestra" # custom description (recommended)
  canonical: "" # custom canonical URL (optional)
  noindex: false # false (default) or true
---

Oakestra supports the deployment of unikernels built using [Unikraft](http://unikraft.org). 

{{< callout context="caution" title="Requirements" icon="outline/alert-triangle">}}

All you need to start deploying unikernels is:

> 1. A unikernel .tar.gz hosted somewhere accessible to the Oakestra.

Reviewer: ```suggestion
1. A unikernel `.tar.gz` hosted somewhere accessible to Oakestra
```

2. A service deployment descriptor 

> 3. A Worker Node with a Unikernel Runtime enabled

Reviewer: ```suggestion
3. A worker node with a unikernel runtime enabled
```

{{< /callout >}}

## Packaging your first Unikraft image 

> Once you build your first Unikernel using Unikraft, you can create a tarball of the kernel image and the necessary files.

Reviewer: ```suggestion
Once you build your first unikernel using Unikraft, you can create a tarball of the kernel image and the necessary files.
```

> ```bash

Reviewer: ```suggestion
```yaml
```

Author: that’s actually the output of the `tree` bash command

myunikernel.tar.gz
|-- kernel
`--files1/
      |--...
 `--...
```

> Your unikernel tarball MUST contain a file named `kernel`, which is the unikernel image itself. Additionally, it CAN contain a `files1` folder, which is mounted by Oakestra as kernel filesystem at runtime. This folder contains all the additional files for the unikernel to run.

Reviewer: ```suggestion
Your unikernel tarball MUST contain a file named `kernel`, which is the unikernel image itself. Additionally, it CAN contain a `files1` folder, which is mounted by Oakestra as a kernel filesystem at runtime. This folder contains all the additional files required by the unikernel to run.
```

 }
 ]
 }
 ```

 #### Note

- The `virtualization` field is set to `unikernel`
- The `code` field contains the URL to the unikernel tarball. This URL must be accessible to the Oakestra worker nodes. We're using the Oakestra release page as an example here to host the tarball file.
- The `arch` field specifies the architecture of the unikernel. In this case, it's `amd64`.

The rest of the fields are similar to the ones used for container services.
You can then deploy the application as usual using the Oakestra Dashboard, the Oakestra CLI, or the Oakestra APIs.

## Enable a KVM based Unikernel Runtime for an Oakestra Worker Node

If your node supports nested virtualization and you have KVM installed, you can enable the KVM runtime for your Oakestra worker node. This will allow you to deploy unikernels on your worker node.

{{< callout context="tip" title="How to install KVM?" icon="outline/rocket" >}}
E.g., If your worker node uses an Ubuntu distro, you can follow [this](https://phoenixnap.com/kb/ubuntu-install-kvm) guide to install KVM.
{{< /callout >}}

Follow these steps on the **worker node**:

1. Stop your worker node (if already running)
```bash
sudo NodeEngine stop
```

> 2: Enable unikernel runtime using 

Reviewer: ```suggestion
2. Enable unikernel runtime using 
```

```bash
sudo NodeEngine config virtualization unikernel on
```

> 3: Re-start your worker node

Reviewer: ```suggestion
3. Re-start your worker node
```

---
title: "Unikernel Deployment"
summary: ""
draft: false
weight: 301
toc: true
seo:
  title: "Unikernel Deployment" # custom title (optional)
  description: "Deployment of Unikraft Unikernels in Oakestra" # custom description (recommended)
  canonical: "" # custom canonical URL (optional)
  noindex: false # false (default) or true
---

Oakestra supports the deployment of unikernels built using [Unikraft](http://unikraft.org). 

> #### What do you need?

Reviewer: I think this should be done in the style of the the other requirements callouts such as on the `oak-environment/high-level-setup-overview/` page

It's simple. All you need to start deploying unikernels is:

1. A unikernel .tar.gz hosted somewhere accessible to the Oakestra.
2. A service deployment descriptor 
3. A worker Node with KVM Runtime enabled

## Packaging your first Unikraft image 

> Once you build your first Unikernel using [Unikraft](http://unikraft.org), you can create a tarball of the kernel image and the necessary files.

Reviewer: ```suggestion
Once you build your first Unikernel using Unikraft, you can create a tarball of the kernel image and the necessary files.
```
Since it's linked a few lines above this is probably a bit superfluous

```bash
myunikernel.tar.gz
|-- kernel
`--files1/
      |--...
 `--...
```

Your unikernel tarball MUST contain a file named `kernel`, which is the unikernel image itself. Additionally, it CAN contain a `files1` folder, which is mounted by Oakestra as kernel filesystem at runtime. This folder contains all the additional files for the unikernel to run.

{{< callout context="tip" title="Looking for an example?" icon="outline/rocket" >}}

> Check out the **Unikraft Nginx** example in the Application Catalogue section of the wiki.

Reviewer: ```suggestion
Check out the **[Unikraft Nginx](../app-catalog/nginx-unikernel-deployment/)** example in the Application Catalogue section of the wiki.
```
This is definitely requires a link, but it won't work until this branch is rebased onto the app-catalog branch

Author: created an issue to fix this when that's merged. Broken links are harder to spot than missing ones :D

Unikernels can be deployed to Oakestra like any other service. You need to create a service deployment descriptor that describes the unikernel service you want to deploy.
Here is an example of an Nginx server using Unikraft:

```json {title="unikernel-nginx.json"}
 {
   "sla_version": "v2.0",
   "customerID": "Admin",
   "applications": [
 {
       "applicationID": "",
       "application_name": "nginx",
       "application_namespace": "test",
       "application_desc": "Simple demo of a Unikraft Nginx server",
       "microservices": [
 {
           "microserviceID": "",
           "microservice_name": "nginx",
           "microservice_namespace": "nginx",
           "virtualization": "unikernel",
           "cmd": [
             ""
 ],
           "memory": 400,
           "vcpus": 1,
           "vgpus": 0,
           "vtpus": 0,
           "bandwidth_in": 0,
           "bandwidth_out": 0,
           "port": "9000:80",
           "storage": 0,
           "code": "https://github.com/oakestra/oakestra/releases/download/alpha-v0.4.301/nginx_amd64.tar.gz",
           "arch": [
             "amd64"
 ],
           "state": "",
           "added_files": []
 }
 ]
 }
 ```

>  #### What's important to note in this SLA?

Reviewer: ```suggestion
 #### Note
```
I think this is a bit long

- The `virtualization` field is set to `unikernel`
- The `code` field contains the URL to the unikernel tarball. This URL must be accessible to the Oakestra worker nodes. We're using the Oakestra release page as an example here to host the tarball file.
- The `arch` field specifies the architecture of the unikernel. In this case, it's `amd64`.

The rest of the fields are similar to the ones used for container services.
You can then deploy the application as usual using the Oakestra Dashboard, the Oakestra CLI, or the Oakestra APIs.

> {{< callout context="caution" title="Caution" icon="outline/alert-triangle" >}}

Reviewer: I think just `Caution` isn't that concise as a title

Reviewer: That being said, if the highlight this requirement at the beginning of the page, it might be a bit superfluous here

> You need at least one worker node in one of your clusters with a unikernel runtime enabled to deploy your unikernels. 

Reviewer: ```suggestion
You need at least one worker node in one of your clusters with a KVM runtime enabled to deploy your unikernels. 
```

{{< /callout >}}

## Enable a KVM Runtime for an Oakestra Worker Node

If your node supports nested virtualization and you have KVM installed, you can enable the KVM runtime for your Oakestra worker node.

{{< callout context="tip" title="How to install KVM?" icon="outline/rocket" >}}

> Please follow [this](https://phoenixnap.com/kb/ubuntu-install-kvm) guide to install KVM on your worker node.

Reviewer: I think we should mention that this is ubuntu specific

Author: ```suggestion
E.g., If your worker node uses an Ubuntu distro, you can follow [this](https://phoenixnap.com/kb/ubuntu-install-kvm) guide to install KVM on your worker node.
```

{{< /callout >}}

> All you need to do is:

Reviewer: ```suggestion
Follow these steps on the **worker node**:
```

> 1: Stop your worker node (if already running)

Reviewer: ```suggestion
1. Stop your worker node (if already running)
```
In other places in the docs we use the simple `.` instead of a colon

Reviewer: ```suggestion
1. Stop the node engine (if already running)
```

```bash
sudo NodeEngine stop
```
2: Enable unikernel runtime using 
```bash
sudo NodeEngine config virtualization unikernel on
```
3: Re-start your worker node
```bash
sudo NodeEngine -a <Cluster Orchestrator IP> -d
```

> {{< callout context="note" title="Note" icon="outline/info-circle" >}}

Reviewer: Maybe you could find a more concise title than `Note`?
